test/test_wan_strategy.py

```diff
@@ -0,0 +1,166 @@
+    """
+    Verify the per-device WAN topology decoding, using a dashed device identifier with too few components.
+    The solution is to pad the segments with `default` labels at the front.
+
+    This specific test uses a vanilla TTN device identifier.
+    """
+    strategy = WanBusStrategy()
+    topology = strategy.topic_to_topology("myrealm/dt/eui-70b3d57ed005dac6/data.json")
+    assert topology == SmartMunch(
+        {
+            "realm": "myrealm",
+            "network": "default",
+            "gateway": "eui",
+            "node": "70b3d57ed005dac6",
+            "slot": "data.json",
+        }
+    )
+
+
+@pytest.mark.strategy
+def test_wan_strategy_device_dashed_topo_three_components():
+    """
+    Verify the per-device WAN topology decoding, using a dashed device identifier with exactly three components.
+    This topic will be decoded as-is.
+    """
+    strategy = WanBusStrategy()
+    topology = strategy.topic_to_topology("myrealm/dt/acme-area42-eui70b3d57ed005dac6/data.json")
+    assert topology == SmartMunch(
+        {
+            "realm": "myrealm",
+            "network": "acme",
+            "gateway": "area42",
+            "node": "eui70b3d57ed005dac6",
+            "slot": "data.json",
+        }
+    )
+
+
+@pytest.mark.strategy
+def test_wan_strategy_device_dashed_topo_too_many_components_merge_suffixes():
+    """
+    Verify the per-device WAN topology decoding, using a dashed device identifier with too many components.
+    The solution is to merge all trailing segments into the `node` slot, re-joining them with `-`.
+    """
+    strategy = WanBusStrategy()
+    topology = strategy.topic_to_topology("myrealm/dt/acme-area42-eui70b3d57ed005dac6-suffix/data.json")
+    assert topology == SmartMunch(
+        {
+            "realm": "myrealm",
+            "network": "acme",
+            "gateway": "area42",
+            "node": "eui70b3d57ed005dac6-suffix",
+            "slot": "data.json",
+        }
+    )
+
+
+@pytest.mark.strategy
+def test_wan_strategy_device_dashed_topo_too_many_components_redundant_realm():
+    """
+    Verify the per-device WAN topology decoding, using a dashed device identifier with too many components.
+
+    This is an edge case where the first addressing component actually equals the realm.
+    Strictly, it is a misconfiguration, but we pretend to be smart, and ignore that,
+    effectively not using the redundant information, and ignoring the `myrealm-` prefix
+    within the device identifier slot altogether.
+
+    Cheers, @thiasB.
+    """
+    strategy = WanBusStrategy()
+    topology = strategy.topic_to_topology("myrealm/dt/myrealm-acme-area42-eui70b3d57ed005dac6/data.json")
+    assert topology == SmartMunch(
+        {
+            "realm": "myrealm",
+            "network": "acme",
+            "gateway": "area42",
+            "node": "eui70b3d57ed005dac6",
+            "slot": "data.json",
+        }
+    )
```

Comment:
@thiasB: When looking at [^1], your devices are probably configured with device identifiers like `device_id=hiveeyes-thias-freiland-hive2`, so they _do_ include the realm at the first device identifier segment. This test case demonstrates that it will be ignored, so that you do not have to re-program the devices.

Did I get this detail right?

[^1]: https://community.hiveeyes.org/t/tts-ttn-daten-an-kotori-weiterleiten/1422/35

Reply:
Yes, you are right. However, re-creating the devices in the TTN console wouldn't take much effort. If the solution below is active, I will be happy to not touch device registration. For new devices, I will follow the rule that the realm is not included in the device ID.